sklearn/multioutput.py

```diff
@@ -359,3 +362,224 @@ def score(self, X, y):
+        If cv is None the true labels are used when fitting. Otherwise
+        possible inputs for cv are:
+            * integer, to specify the number of folds in a (Stratified)KFold,
+            * An object to be used as a cross-validation generator.
+            * An iterable yielding train, test splits.
+
+    random_state : int, RandomState instance or None, optional (default=None)
+        If int, random_state is the seed used by the random number generator;
+        If RandomState instance, random_state is the random number generator;
+        If None, the random number generator is the RandomState instance used
+        by `np.random`.
+
+        Used to generate random chain orders.
+
+    Attributes
+    ----------
+    classes_ : list of arrays of length len(estimators_) containing the
+        class labels for each estimator in the chain.
+
+    estimators_ : list
+        A list of clones of base_estimator. Once fit the estimators in
+        this list will be ordered as specified by the order attribute.
+        The original order of labels (as specified by Y) is recovered in
+        the predict method by indexing into the predictions with the
+        list of indices in the order attribute.
+
+    References
+    ----------
+    Jesse Read, Bernhard Pfahringer, Geoff Holmes, Eibe Frank, "Classifier
+    Chains for Multi-label Classification", 2009.
+
+    """
+    def __init__(self, base_estimator, order=None, cv=None, random_state=None):
+        self.base_estimator = base_estimator
+        self.order = order
+        self.cv = cv
+        self.random_state = random_state
+
+    def fit(self, X, Y):
+        """Fit the model to data matrix X and targets Y.
+
+        Parameters
+        ----------
+        X : {array-like, sparse matrix}, shape (n_samples, n_features)
+        Y : {array-like, sparse matrix}, shape (n_samples, n_classes)
+
+        Returns
+        -------
+        self : object
+            Returns self.
+        """
+        X, Y = check_X_y(X, Y,
+                         multi_output=True,
+                         accept_sparse=True)
+
+        random_state = check_random_state(self.random_state)
+
+        self.order_ = self.order
+        if self.order_ is None:
+            self.order_ = np.array(range(Y.shape[1]))
+        elif isinstance(self.order_, str):
+            if self.order_ == 'random':
+                self.order_ = random_state.permutation(Y.shape[1])
+        elif sorted(self.order_) != list(range(Y.shape[1])):
+                raise ValueError("invalid order")
+
+        self.estimators_ = [clone(self.base_estimator)
+                            for _ in range(Y.shape[1])]
+
+        self.classes_ = []
+
+        if self.cv is None:
+            Y_pred_chain = Y[:, self.order_]
+        else:
+            Y_pred_chain = np.zeros((X.shape[0], len(self.estimators_)))
+
+        if sp.issparse(X):
+            X_aug = sp.hstack((X, Y_pred_chain), format='lil')
```

Comment:
I get why lil is good (sort of) from the perspective of inserting successive predictions. But we usually benefit from CSR in predictors. In the cv is None case, therefore, CSR is much preferred. In the other case, we could do CSR too, but efficient insertion into an empty column takes a few lines of code; CSC would be a little more appropriate. You needn't worry about that case for now.